mesonbuild/msetup.py

```diff
@@ -246,7 +253,10 @@ def _generate(self, env: environment.Environment) -> None:
                 fname = os.path.join(self.build_dir, 'meson-logs', fname)
                 profile.runctx('intr.backend.generate()', globals(), locals(), filename=fname)
             else:
-                intr.backend.generate()
+                captured_compile_args = intr.backend.generate(
```

Comment:
I noticed this while looking at my other thought for reducing the effect on the function signature for generate():

The shower thought that in theory one could save the vslite context as an instance variable and retrieve it from intr.backend after invoking intr.backend.generate. I haven't taken a close enough look to see which API is better, especially since it's a moot point at the moment... I'm not positive it's better because it might require *paired* functions at a minimum, one to set it and one to retrieve it. How elegant is that? 🤔

Reply:
I did review but I mostly go by tests. It is just plain impossible to consider all the ways things can fail. At the very least this has shown that there is no test for profile-self.

Since this MR seems to have some issues unless they get fixed in time for the release I'll revert this out. Sorry.

Reply:
> This is broken if profile-self is used...

I'm not familiar with this profile path.  I guess this isn't covered by any of the build check automation.  Are there instructions for what we need to do to test this?  Anything beyond just adding `--profile-self` in a test and ensuring succeeds (and in this case, fixing up the 'statement' arg to `profile.runctx(...)` to now match the new signature)?

> The shower thought that in theory one could save the vslite context as an instance variable and retrieve it from intr.backend after invoking intr.backend.generate.

I also considered this kind of approach but wasn't keen on it (at least just yet) for a few reasons -
- If in doubt, I'll lean towards clarity, and strict, explicit code (params/state), over hiding params/state/dependencies/functionality/etc.
- Although it would make the generate() interface a little cleaner, I _think_ the total amount of additional code surface area would be more overall.
- I think, if `intr.backend` now has to carry around some extra member state (flags to indicate we want it to provide the captured compile args when we call 'generate' and then extra member state to hold on to those results, post-generate, for when we _might_ then request them), then the lifetime of that data is now much more sloppy and spread out than it is when we have a nice, tight, and strict window in which all this generate context/state is in play.
- and less strict aspect of an interface would need a bit more validation to sanity check no one's trying to get hold of these results before they've been generated, and things like that.
- This would appear to be the first feature of meson (that I'm aware of) that needs to pull together results from multiple setup stages, so, at this time, it feels like it may be prematurely constructing some abstraction machinery, based on just one use case, perhaps in anticipation of some unknown future feature that might want other context data, on top of the captured compile args.  Maybe someone will want to add something similar that needs slightly different data from multiple different backend 'generations' (e.g. some other IDE generator that simply glues together the ninja backends but also with access to more details on the individual targets... for whatever reason), at which point perhaps it might look better to hide away a growing list of generate flags/args and resultant data, accessed through a retrieval func... or it might still be OK to simply pass in to generate a structure that aggregates more options for returning captured state from 'generate' which returns a union type of different results depending on the options.

Reply:
> At the very least this has shown that there is no test for profile-self.

We do test profile-self in general in AllPlatformTests.test_command_line, due to the default None value the effect of running the profile-self codepath is that no context is passed around, which is also what running non-vslite backends do.

So my assumption is that running `meson setup --genvslite --profile-self builddir` is broken because it passes around a null context, but needs it to contain something -- and everything inside Vs2010Backend(gen_lite=True) that accesses that dict is going to crash with various AttributeErrors when trying to treat a NoneType as a `list[str]`. We don't specifically check the combination of both, because genvslite is only tested by running `WindowsTests.test_genvslite`.

> Anything beyond just adding `--profile-self` in a test and ensuring succeeds (and in this case, fixing up the 'statement' arg to `profile.runctx(...)` to now match the new signature)?

Probably this is sufficient to fix that issue.